src/passes/MinimizeRecGroups.cpp

```diff
@@ -0,0 +1,748 @@
+    groupShapeIndices.erase(it);
+
+    // Put the types for both groups into the canonical order.
+    auto permutation = getCanonicalPermutation(groupInfo.group);
+    groupInfo.permutation = otherInfo.permutation = std::move(permutation);
+
+    // Set up `other` to be the representative element of the equivalence class.
+    otherInfo.classInfo.emplace(otherInfo);
+    auto& classInfo = *otherInfo.classInfo;
+
+    // Update both groups to have the initial valid order. Their shapes still
+    // match.
+    classInfo.permute(otherInfo);
+    classInfo.permute(groupInfo);
+
+    // Insert `other` with its new shape. It may end up being joined to another
+    // existing equivalence class, in which case its class info will be cleared
+    // and `group` will subsequently be added to that same existing class, or
+    // alternatively it may be inserted as the representative element of a new
+    // class to which `group` will subsequently be joined.
+    shapesToUpdate.push_back(group);
+    shapesToUpdate.push_back(other);
+  }
+
+  std::vector<size_t>
+  getCanonicalPermutation(const std::vector<HeapType>& types) {
+    // The correctness of this part depends on some interesting properties of
+    // strongly connected graphs with ordered, directed edges. A permutation of
+    // the vertices in a graph is an isomorphism that produces an isomorphic
+    // graph. An automorphism is an isomorphism of a structure onto itself, so
+    // an automorphism of a graph is a permutation of the vertices that does not
+    // change the graph. Permutations can be described in terms of sets of
+    // cycles of elements.
+    //
+    // Theorem 1: All cycles in an automorphism of an SCC are the same size.
+    //
+    //     Proof: By contradiction. Assume there are two cycles of different
+    //     sizes. Because the SCC is fully connected, there is a path from an
+    //     element in the smaller of these cycles to an element in the larger.
+    //     This path must contain an edge between two vertices in cycles of
+    //     different sizes N and M such that N < M. Apply the automorphism N
+    //     times. The source of this edge has been cycled back to its original
+    //     index, but the destination is not yet back at its original index, so
+    //     the edge's destination index is different from its original
+    //     destination index and the permutation we applied was not an
+    //     automorphism after all.
+    //
+    // Corollary 1.1: No nontrivial automorphism of an SCC may have a stationary
+    // element, since either all the cycles have size 1 and the automorphism is
+    // trivial or all the cycles have some other size and there are no
+    // stationary elements.
+    //
+    // Corollary 1.2: No two distinct isomorphisms of an SCC with the same first
+    // element are automorphic to each other since no nontrivial automorphism
+    // can keep the first element stationary.
```

Comment:
Probably related to my confusion about the term "automorphic": I don't follow this. Before, we defined it like this:

> an automorphism of a graph is a permutation of the vertices that does not change the graph

So if two SCCs are distinct, then they are not the same graph. So no automorphism is even relevant to speak of here: any mapping between them changes the graph, so it cannot be an automorphism, so no automorphism can exist. That is, we don't need any facts about the first element. So I must be missing the point yet again :smile:

Reply:
Returning to our example:
```
(rec
  (type $a1 (struct (field (ref $a2) (ref $b1))))
  (type $b1 (struct (field (ref $b2) (ref $a1))))
  (type $a2 (struct (field (ref $a1) (ref $b2))))
  (type $b2 (struct (field (ref $b1) (ref $a2))))
)
```
`A: $a1, $b1, $a2, $b2` and `B: $a1, $a2, $b1, $b2` are distinct permutations (i.e. distinct isomorphisms) of this SCC. They have the same first element, `$a1`. The corollary says that there is no automorphism of `A` onto `B` or vice versa (i.e. `A` and `B` are not "automorphic" by analogy to the relationship between "isomorphism" and "isomorphic").

Reply:
Hmm, I really think we should think about another term here:

> The corollary says that there is no automorphism of `A` onto `B`

Automorphisms are of a thing on itself. That is, it is a function from `X` to `X`, not `X` to `Y`. That's wikipedia's definition and the one I've always heard. And we were consistent with it before, when we said "here is a rec group Foo and here is an automorphism on it in cycle notation".

Also, any automorphism (in that old sense) on `A` will turn `A` into itself, so no such transform can get to `B` (since `A != B`). Likewise any automorphism on `B` cannot get to `A`.

OTOH there *is* a transform that gets from `A` to `B`: `0,2,1,3`. So my best guess is that you want to disallow this particular transform but allow others, but I can't see how/why yet.

Reply:
`$a1, $b1, $a2, $b2` has an automorphism `($a1 $a2) ($b1 $b2)`, so it is automorphic to `$a2, $b2, $a1, $b1` even though those are distinct permutations that have different initial elements. If you write down the vertices of these graphs (i.e. colors and ordered edges), you will write down the same graph twice. But `$a1` is a different type from `$a2` even though they are automorphic to each other.

Reply:
After offline discussion, it looks like the best path forward is to clarify that this corollary is using a notion of type identity that is not related to the structures of emitted rec groups. In general, it would be good to define all the different equivalence relations we are using up front.